Versioned protobuf files

mzabaluev opened the pull request:
Resolves #94
Delivers proto and code generation changes for #95

Changes broken down for reviewers:

* fdcca7813a7676791163c44f88edd9931a6f3f2a - Adds `cometbft.*` versioned proto files with evolution going back to v0.34.x.
* 784f976dc46f9a4d4f24133605051f06b344601a - Relocates and adds regenerated `*.pb.go`. The bulk of changes is generated code.
* 5a5beff861b509fcc3a891756891d43ef88e6858 - Moves and adapts support files adding methods alongside `*.pb.go` types. Defines aliases to latest types in unversioned domain packages to ease importing.
* ba3c57430476773ca340541199eab0b70784d23f - Fixes proto imports and aliases in production code.
* 9cbf604994cb79a8454cdf78e6019cb8cc575de2 - Fixes proto imports and aliases in tests and examples.

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

mzabaluev commented on Makefile:
```diff
@@ -167,8 +167,8 @@ endif
 	@echo "Generating Protobuf files"
 	@go run github.com/bufbuild/buf/cmd/buf generate
-	@mv ./proto/tendermint/abci/types.pb.go ./abci/types/
-	@cp ./proto/tendermint/rpc/grpc/types.pb.go ./rpc/grpc
```
No idea why this generated Go file had to be copied while the one for `abci` above is moved. @julienbrt @thanethomson can you clarify why this duplication was needed in https://github.com/tendermint/tendermint/pull/9356?

thanethomson replied:
This is a legacy thing, it seems as though it was deemed best to have the generated ABCI proto code colocated with the rest of the ABCI code.

Similarly with the generated gRPC code.

The better approach is to have it all generated into an `/api/` package in the root of the project, as per Buf's guidelines.

thanethomson commented:
Is the intention here to eventually have all generated Go code in the `/api/` package, as discussed in #95?

mzabaluev commented on privval/file.go:
```diff
@@ -16,7 +16,8 @@ import (
 	"github.com/cometbft/cometbft/libs/protoio"
 	"github.com/cometbft/cometbft/libs/tempfile"
-	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	cmtproto "github.com/cometbft/cometbft/proto/cometbft/types/v3"
+	cmtproto1 "github.com/cometbft/cometbft/proto/cometbft/types/v1"
```
The dispersal of messages between differently versioned packages is annoying and makes the generated packages hard to use. The reason to do it this way was to avoid duplication in message definitions and generated code. But perhaps it's better when all message definitions for an API domain are present in each version (except if they are deprecated), so that e.g. you get everything you need from types by importing `cometbft/types/v3`.

mzabaluev commented on proto/cometbft/abci/v3/types.proto:
```diff
@@ -0,0 +1,220 @@
+// https://github.com/cosmos/gogoproto/blob/master/custom_types.md#warnings-and-issues
+
+service ABCI {
+  rpc Echo(v1.RequestEcho) returns (v1.ResponseEcho);
```
I think there is a kludgy solution to this in Go.

Alongside the package structure for generated `*.pb.go` files, we can add in manually written Go files aliasing the names from the versioned packages into versionless package paths denoting "the latest API". So e.g. `Message` in `cometbft/blocksync/types.go` will be the alias of `v2.Message` as generated into `cometbft/blocksync/v2/types.pb.go`. I need to look into if the whole content of the future `./api` module could be such aliases, while the generated versioned packages will be put under  `./proto/gen` or elsewhere, and not meant to be used directly as public API.

One tedious thing with this approach that I don't see how to work around is aliasing constants such as [enum values](https://github.com/cometbft/cometbft/pull/495/files#diff-a2793ccde5e15af24d28f6c1af58cb3a60625f9e2c43bfd12b2d893c37744571R114).

sergio-mena commented:
This PR is gigantic. Is there are way structure it to make it easier to review (maybe rearranging the contents of the commits?). A review on this kind of PRs cannot be more than "best-effort" from the reviewer, which may compromise quality.

mzabaluev commented:
@sergio-mena I would not recommend reviewing individual commits, as they have a history of interim changes and many files were moved, or removed and restored in a different place, in different commits.

For convenience, I could split this into a PR just adding the new protobuf file structure without touching code generation, and a PR changing the affected Go code in one neatly squashed change set. I don't think fragmenting it further would make sense, as the code changes are pretty similar across the code base and it's mostly trivial renaming and relocation.
You can pretty much skip any generated `*.pb.go` files, too.

This is not how this eventual state could be reached, though, as I worked through a few mistakes in versioned protos that were revealed by errors in the Go build, and addressed the issues with added-on Go files defining methods on generated types.

sergio-mena commented:
> just adding the new protobuf file structure without touching code generation, and a PR changing the affected Go code in one neatly squashed change set.

Thanks for proposing this, I think it would definitely help. A few more comments:

* Up to you, but IMO you don't need to split this across different PRs, it would be enough to rearrange your commits to split the different parts that could be reviewed in isolation. The advantage you have with rearranging commits is that they don't need to build or pass the tests individually (which is a requirement for individual PRs)
* Please give it a thought on whether you can go further and provide a finer-grained set of changes. I know we haven't discussed yet about concrete limits in our team. But I'd say that, in general, anything above 2000-3000 changed lines (total, + and -) in a single PR (or a single reviewable commit) is too big.

thanethomson commented on abci/types/types.go:
```diff
@@ -1,125 +1,169 @@
+type RequestFinalizeBlock = v3.RequestFinalizeBlock
+
+// Discriminated Request variants are defined in the latest proto package.
+type Request_Echo = v3.Request_Echo
```
To resolve the linter errors, I'd recommend either doing a `//nolint:revive` right at the top of the file to cover the whole file, or to update the [linter settings](https://github.com/cometbft/cometbft/blob/main/.golangci.yml) to ignore this file. I think turning off the specific lint is probably best.

thanethomson commented on api/cometbft/blocksync/types.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+type Message = v2.Message
+type Message_BlockRequest = v2.Message_BlockRequest
```
Same as the previous comment re: the lints.

sergio-mena commented on proto/cometbft/state/v2/types.proto:
```diff
@@ -0,0 +1,73 @@
+// It is persisted to disk for each height before calling Commit.
+message ABCIResponses {
+  repeated cometbft.abci.v2.ResponseDeliverTx deliver_txs = 1;
+  cometbft.abci.v2.ResponseEndBlock end_block = 2;
+  cometbft.abci.v2.ResponseBeginBlock begin_block = 3;
```
These might disappear (see my comments above), which may cause this one to disappear as well

sergio-mena commented on proto/cometbft/types/v1/block.proto:
```diff
@@ -0,0 +1,15 @@
+message Block {
+  Header                        header      = 1 [(gogoproto.nullable) = false];
+  Data                          data        = 2 [(gogoproto.nullable) = false];
+  cometbft.types.v1.EvidenceList evidence    = 3 [(gogoproto.nullable) = false];
```
Indentation?

sergio-mena reviewed (commented):
This is a review of the commit containing the changes to protos

sergio-mena reviewed (commented):
This is a review for the remaining commits

thanethomson commented on proto/cometbft/abci/v1/types.proto:
```diff
@@ -0,0 +1,409 @@
+// ConsensusParams contains all consensus-relevant parameters
+// that can be adjusted by the abci app
+message ConsensusParams {
+  BlockParams                      block     = 1;
```
Run `make proto-format` for this. You'll need `clang-format` installed locally first.

mzabaluev replied:
Done in https://github.com/cometbft/cometbft/pull/495/commits/cc351d526eecf0a888d5ca429381efad40b9cbc0

thanethomson commented on proto/cometbft/abci/v2/types.proto:
```diff
@@ -0,0 +1,278 @@
+  bytes                             app_hash         = 3;
+}
+
+message ResponseBeginBlock {
```
I think because the `Event` type changed. Changes in types upon which a type is dependent has a "viral" effect in triggering the need for updates, so if the `Event` type changes in some breaking way, all types that depend on it need new versions.

If the `Event` type changes in a non-breaking way, or doesn't change, then this viral effect isn't present.

mzabaluev replied:
Yes, it's the new `Event`.

mzabaluev commented on proto/cometbft/abci/v3/types.proto:
```diff
@@ -0,0 +1,288 @@
+  int64                            initial_height   = 6;
+}
+
+message RequestPrepareProposal {
```
`ExtendedCommitInfo` has changed content, precipitating the new version here as well.

mzabaluev commented on proto/cometbft/abci/v3/types.proto:
```diff
@@ -0,0 +1,288 @@
+  bytes proposer_address = 8;
+}
+
+message RequestProcessProposal {
```
This is because of `CommitInfo`.

Perhaps it's not easy to see at a glance when the field types are referenced from the same package. So it was `v2.CommitInfo` for `v2.RequestProcessProposal`, but in the proto it looks as though the field definition has not changed in `v3`. Maybe I should disambiguate all type references with the v-package prefix for clarity, even where it refers to the file's own package.

mzabaluev commented on proto/cometbft/blockchain/v1/types.proto:
```diff
@@ -0,0 +1,41 @@
+syntax = "proto3";
```
I think it's fine to erase this and tell people to just rename `tendermint.blockchain` to `cometbft.blocksync.v1` (or do the same in their target language equivalent) in migration notes.

mzabaluev replied:
Right, removed `blockchain/v1` in https://github.com/cometbft/cometbft/pull/495/commits/2b47221a98922b40f675926aa59336c5ee9f6f51

mzabaluev commented on proto/cometbft/consensus/v2/wal.proto:
```diff
@@ -0,0 +1,33 @@
+import "google/protobuf/timestamp.proto";
+
+// MsgInfo are msgs from the reactor which may update the state
+message MsgInfo {
```
No, it contains a `v2.Message` field.

mzabaluev commented on proto/cometbft/types/v2/params.proto:
```diff
@@ -0,0 +1,39 @@
+  option (gogoproto.equal)    = true;
+
+  // Was named app_version in v1
+  uint64 app = 1;
```
I'm not opposed to renamings for cosmetic reasons, and "version.app_version" was redundant.

But now that we are effectively introducing a whole new namespace of protos and their generated bindings, we could apply this renaming retroactively and tell tendermint -> CometBFT 0.34 migrants to use `app` as the field name.

> This caused the need for new versions of `ConsensusParams`

Not only this, though, there was also a field in `BlockParams` that got reserved in `v2`.

mzabaluev replied:
Erased `v2.VersionParams` and renamed the `v1.VersionParams` field to `app` in https://github.com/cometbft/cometbft/pull/495/commits/d2c36a97b35c5c68509d5d874d3bd8be262c5b07

mzabaluev commented on types/signed_msg_type.go:
```diff
@@ -1,11 +1,20 @@
+	SignedMsgType_PREVOTE   SignedMsgType = cmtproto.PrevoteType
+	SignedMsgType_PRECOMMIT SignedMsgType = cmtproto.PrecommitType
+	SignedMsgType_PROPOSAL  SignedMsgType = cmtproto.ProposalType
+)
```
This is some activism on my part. Those const names differ from the naming convention used by gogoproto for other enums by default, and they are also poorly namespaced: it's not clear at a glance what "UnknownType" refers to, among the dozens of names that are defined in `types`. So I brought them in line with the general convention, even though the proto source gives the name overrides specifically for Go. I can revert these to the previously established short names, since everybody should have gotten used to them by now.

mzabaluev replied:
Reverted in https://github.com/cometbft/cometbft/pull/495/commits/57e8a818633132ac475f5a7bbb8302b3b0b9766f

mzabaluev commented on proto/cometbft/abci/v2/types.proto:
```diff
@@ -0,0 +1,278 @@
+      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+}
+
+message ResponseCheckTx {
```
This uses `v2.Event` instead of `v1.Event`.

mzabaluev commented on proto/cometbft/abci/v2/types.proto:
```diff
@@ -0,0 +1,278 @@
+  string mempool_error = 11;
+}
+
+message ResponseDeliverTx {
```
Same as above.

sergio-mena reviewed (approved):
LGTM, thanks for your efforts

sergio-mena commented:
One last thing from my side, I'm reading [this](https://protobuf.dev/programming-guides/dos-donts/), and I think it would be a good moment to rename all \*__UNKNOWN_ enums we have to \*__UNSPECIFIED_, to align with the recommendations.
Also, please check if we have an enum without a UNKNOWN/UNSPECIFIED zero-value

mzabaluev commented:
> One last thing from my side. I'm reading [this](https://protobuf.dev/programming-guides/dos-donts/), and I think it would be a good moment to rename all *__UNKNOWN_ enums we have to *__UNSPECIFIED_, to align with the recommendations. Also, please check if we have an enum without an UNKNOWN/UNSPECIFIED zero-value

Since we are on a feature branch, I will take this out into a follow-on PR.
